File: services/experimentation-portal/src/api/experiments/model/experimentResults.ts
import type { Metric } from "../../metricsCatalog";
import type {
  CreateExperimentMetricDirection,
  CreateExperimentMetricRole,
} from "./experiment";

export type DecisionRecommendation =
  | "DECISION_RECOMMENDATION_UNSPECIFIED"
  | "DECISION_RECOMMENDATION_RECOMMEND"
  | "DECISION_RECOMMENDATION_NOT_RECOMMEND"
  | "DECISION_RECOMMENDATION_INCONCLUSIVE";

export type ZTestResult = {
  absolute_difference: boolean;
  ci_lower: number;
  ci_upper: number;
  p_value: number;
  adjusted_ci_lower: number;
  adjusted_ci_upper: number;
  adjusted_p_value: number;
  is_significant: boolean;
  powered_effect: number;
};

export type MetricValue = {
  numerator: number;
  denominator: number;
};

export type EnrichedExperimentMetric = {
  metric_id: Metric;
  type: CreateExperimentMetricRole;
  direction: CreateExperimentMetricDirection;
  mde?: number;
  nim?: number;
};

export type ExperimentResultsResponse = {
  decision_recommendation: DecisionRecommendation;
  recommendation_reason: string;
  practically_significant: boolean;
  statistically_significant: boolean;
  test_results: Record<string, ZTestResult>;
  metrics: Record<string, EnrichedExperimentMetric>;
  metric_observations: Record<string, Record<string, MetricValue>>;
};

File: libs/prismdbmodels/model/experimentResults/experimentResults.go
package experimentResults

import (
	"github.com/Dan-Sones/prismdbmodels/model/metric"
	"github.com/google/uuid"
)

// How we map metrics to results here is going to be intresting
// a t test and z test might be structurally different
// however we still want them to come down mapped against metric ID
// or maybe we have a separate map for different types of tests?
// maybe a use for generics?

type ExperimentResults struct {
	DecisionRecommendation DecisionRecommendation               `json:"decision_recommendation"`
	RecommendationReason   string                               `json:"recommendation_reason"`
	TestResults            map[uuid.UUID]ZTestResult            `json:"test_results"`
	Metrics                map[uuid.UUID]metric.Metric          `json:"metrics"`
	MetricValues           map[uuid.UUID]map[string]MetricValue `json:"metric_values"`
}

type ZTestResult struct {
	AbsoluteDifference bool    `json:"absolute_difference"`
	CILower            float64 `json:"ci_lower"`
	CIUpper            float64 `json:"ci_upper"`
	PValue             float64 `json:"p_value"`
	AdjustedCILower    float64 `json:"adjusted_ci_lower"`
	AdjustedCIUpper    float64 `json:"adjusted_ci_upper"`
	AdjustedPValue     float64 `json:"adjusted_p_value"`
	IsSignificant      bool    `json:"is_significant"`
	PoweredEffect      float64 `json:"powered_effect"`
}

type MetricValue struct {
	Numerator   interface{} `json:"numerator"`
	Denominator interface{} `json:"denominator"`
}
